### DISCUSSION TITLE
ES9 conversion and validation

### DISCUSSION THREAD

**Discussion by coclav** [maxGraph/maxGraph#18] [Development]
@mcyph has done an amazing work in converting the project to ES9.

Now, this is a massive piece of work that we need to validate. Lots of people are using mxGraph in its current version (23k download per week on NPM) so we need to make sure this continues to operate 

What are your suggestions? There is no tests in the code base today so we can't really automate that.

I'd suggest 
* make sure all the demos content are converted and manually tested.
* make sure the documentation is up to date on how to upgrade
* get some "real" users to validate the migration (I volunteer for my organisation :) )

Any other ideas?

**Comment by mcyph** [comment]
I don't know what the best course of action is, but my thoughts are, referencing the stated goals on the front page:

> The first priority is to maintain a working version of mxGraph and its npm package

probably can't be achieved at the same time as

> The ambitious stretch goal is to refactor the codebase to create a modern modular, tree shakable, version of mxGraph to reduce the whole package size.

The main reason for this is architectural: the `mxGraph` class acts as a gatekeeper for a large percentage of the other classes, imports a majority of the codebase by default, and (for better or worse) resembles a [God object](https://en.wikipedia.org/wiki/God_object). We could refactor this to move many of the methods to other places, but existing code relies substantially on this class.

At the moment, I think things are at a crossroads: the ES9 conversion is mostly compatible with existing code and I don't have many more "cosmetic" changes planned now things like converting to classes and adding TypeScript to the main parts of the library are mostly complete. Many of the kinds of things that could be done are things like making many of the static/utility classes more stateful. This could allow easier building on the library and components in the future, and reduce bugs by convention (by following SOLID principles and removing `if (... != null)` checks to make things "fail fast") but this essentially would be about introducing bugs and breaking things in the short term with the aim of having less bugs and lowering the barrier to entry in the future.

In addition to the [suggestions](https://github.com/jsGraph/mxgraph/pulls) I made at the PR, testing on applications in the wild could be a good idea, but I think we should ask what our goals are first. The bottom line is the reference implementation is what applications are coded against, so perhaps it could be better to cosmetically revise the existing code using automated methods, and simply commit to fixing browser-related problems as they appear, introducing a feature freeze and developing a version with more radical changes in a separate repo? The options as I see them are:

* Keep maintaining the existing (previous prototype-based) mxGraph by running ESLint and possibly integrating typed-mxgraph docs (but not types), testing periodically against browsers, while developing a "mxGraph-next" project and documenting how to migrate and differences
* Fork the current state of the ES9 code, introducing a feature freeze and testing against as many examples+code in the wild as possible to make sure it behaves like the original version, while maintaining "mxGraph-next"
* Keep extending the current ES9 code to add more features. There will be a risk of breaking existing code with new releases, which means this would need to be heavily test-driven and development would be slower, and have more limited goals within the current architecture

**Reply by mcyph** [reply]
After considering this for a while, my preference is currently to go somewhere between the 1st and 2nd option. The current release is more stable because it's what's being used, and there isn't a good reason I can see to update the npm package with the ES9 version for the time being.

That said, the ES9 version is easier to read and maintain, so it would be nice to make it as stable as reasonably possible+working as an alternative, so that when browser support for the current release becomes a problem we could switch to the "ES9 but compatible" branch and be able to make the modifications quickly as needed.

Does anyone have any thoughts/alternative options I haven't considered??

**Reply by junsikshim** [reply]
The last release was 4.2.2 and I think 4.x should end there. We are not removing anything from the npm, I don't think we are causing too much harm leaving it as is.

The change we are currently making(resolving circular dependencies, enable tree-shaking, and so on) is definitely breaking ones regardless of how much users have to adjust which means it should be 5.0. So, I think we can think a little bit less about keeping backward compatibility and focus more on various enhancements to the point where it's still possible to migrate.

**Reply by coclav** [reply]
I think the licensing issue i'm referring to in #23 is solving the question for us, 
* mxGraph stays as it is (4.x)
* this project is a new beginning
